src/lib/comms/sol-socket-impl-linux.c

```diff
@@ -33,42 +33,36 @@
 struct sol_socket_linux {
     struct sol_socket base;
 
+    bool (*on_can_read)(void *data, struct sol_socket *s);
+    bool (*on_can_write)(void *data, struct sol_socket *s);
+    struct sol_fd *watch;
+    const void *data;
     int fd;
-    struct {
-        bool (*cb)(void *data, struct sol_socket *s);
-        const void *data;
-        struct sol_fd *watch;
-    } read, write;
+    bool read_monitor;
+    bool write_monitor;
```

Comment:
I guess the code would be simpler if instead of separate booleans for each kind of monitor you had a `flags` field.